lib/pgp.ml

```diff
@@ -537,24 +540,216 @@ module Packet = struct
       decode_convention public_key secret_packet convention
   end
 
+  module Signature = struct
+    module Subpacket = struct
+      type key_flag =
+        | Certification
+        | Sign_data
+        | Encrypt_communication
+        | Encrypt_storage
+        | Split
+        | Authentication
+        | Private_partaged
+        | Unknown_flag
+      [@@deriving ord, eq, show]
+
+      let tag_to_keyflag tag =
+        match tag with
+        | 0 -> Certification
+        | 1 -> Sign_data
+        | 2 -> Encrypt_communication
+        | 3 -> Encrypt_storage
+        | 4 -> Split
+        | 5 -> Authentication
+        | 7 -> Private_partaged
+        | _ -> Unknown_flag
+
+      let keyflag_to_string flag =
+        match flag with
+        | Certification -> "This key may be used to certify other keys."
+        | Sign_data -> "This key may be used to sign data."
+        | Encrypt_communication ->
+          "This key may be used to encrypt communications."
+        | Encrypt_storage -> "This key may be used to encrypt storage."
+        | Split ->
+          "The private component of this key may have been split by a \
+           secret-sharing mechanism."
+        | Authentication -> "This key may be used for authentication."
+        | Private_partaged ->
+          "The private component of this key may be in the possession of more \
+           than one person."
+        | Unknown_flag -> "Unknown key flag"
+
+      type t =
+        | Key_expiration_time of int32
+        | Revocation_reason of string
+        | Issuer_id of string
+        | Uses of key_flag list
+        | Unknown
+      [@@deriving ord, eq, show]
+
+      let get_length cs =
+        let first_octet = Cstruct.get_uint8 cs 0 in
+        if first_octet < 192 then
+          (1, first_octet)
+        else if first_octet < 255 then
+          let second_octet = Cstruct.get_uint8 cs 1 in
+          let length = 192 + second_octet + (256 * (first_octet - 192)) in
+          (2, length)
+        else
+          let length = Cstruct.BE.get_uint32 cs 1 in
+          (5, Int32.to_int length)
+
+      let rec keep_some opt_list =
+        match opt_list with
+        | Some x :: _list -> x :: keep_some _list
+        | None :: _list -> keep_some _list
+        | [] -> []
+
+      let decode_uses cs =
+        let nth_bit x n = x land (1 lsl n) <> 0 in
+        let flag_int = Cstruct.get_uint8 cs 1 in
+        let map_tag tag flag =
+          if flag then
+            Some (tag_to_keyflag tag)
+          else
+            None
+        in
+        let flags = List.init 8 (nth_bit flag_int) in
+        if List.nth flags 6 then
+          Error (Printf.sprintf "Bad tag for a key flag: 0x40")
+        else
+          Ok (Uses (keep_some (List.mapi map_tag flags)))
+
+      let decode_code_reason code =
+        match code with
+        | 0 -> Ok ""
```

Comment:
We could if you really think it's better but, in this case, I don't see the point.
The reason why the key is revoked is already a string, with a code (an int) adding some information, this function is just decoding the code in order to add this information to the string.

If we created a variant, I'd create a function int -> variant, a function variant -> string and only use this variant when decoding the code, instantly processing it into a string and forgetting about it. I don't think there is a point of keeping it inside the  type as a variant instead of a string in this case ?

Reply:
Why do say that you would have to convert it back to a string? Why not make it something like:
```
type revocation_reason =
  | Key_superseeded of string
...
```

Reply:
I think it makes it harder (not harder,but makes it it take more time/lines of codes to use) for no real utility, seeing the information it passes.
Well, I guess it allows to not use it for example ?

I'll trust you.